### DISCUSSION TITLE
GC9A01 - Display Disturbances After Initialization

### DISCUSSION THREAD

**Discussion by HerrSchwarz617** [Bodmer/TFT_eSPI#3065] [New processor/display support or capability request]
> _I really wasn't sure if such a small thing really deserved a pull request. Please correct me if this was actually the right way._

![Disturbances](https://github.com/Bodmer/TFT_eSPI/assets/154573235/747dc382-50ba-4a28-960f-5184990d446f)

# Problem Description
When a **GC9A01** display is just connected to the power supply and has just been initialized by **TFT_eSPI**, a visible disturbance can be seen in many colors on the screen. This fault image will light up very briefly. It does not disturb normal operation, but with multiple screens and in a dark environment it is simply not a nice looking thing.

# Problem Localization
Tests have shown that this fault is triggered **after** initializing the display (`TFT_eSPI/TFT_Drivers/GC9A01_Init.h`), because if you add a `delay(5000);` after the very last command `0x29`, this fault remains clearly visible for 5 seconds and before that command the screen remains black for 5 seconds before the fault image light up very briefly.

The command `0x29` means `GC9A01A_DISPLAY_ON`. It switches the display on directly after writing the settings into the display memory. But in this case **without executing a clear** before. An uncleaned random render buffer is therefore displayed.

# Solution
Before switching on the display with the very last command `0x29`, the screen buffer needs to be cleaned.

# Fix
> There might be a better solution than using the existing `fillScreen` function with end and begin tft_write, but it works and that was important to me.

After this fix, the disturbance was no more displayed once.
```diff
--- TFT_eSPI/TFT_Drivers/GC9A01_Init.h 2023-12-22 03:52:29.771452500 +0100
+++ TFT_eSPI/TFT_Drivers/GC9A01_Init.h 2023-12-22 03:57:40.179250100 +0100
@@ -227,6 +227,16 @@

   writecommand(0x11);
   delay(120);
+
+  // clean render buffer
+  end_tft_write();
+  fillScreen(TFT_BLACK);
+  begin_tft_write();
+
+  // turn on the screen
   writecommand(0x29);
   delay(20);
 }
```

**Comment by Bodmer** [comment]
Thanks for raising this. Not all displays have this problem, I will try this. Another solution for some displays is to keep the backlight off until the display is initialised.

**Reply by xsong21** [reply]
Thank you for the suggestion. Turning on the backlight after initializing the display and UI can solve this issue.

``` C
// Turn off the backlight 
digitalWrite(DISPLAY_BACKLIGHT_PIN, LOW); 
// Initialize the screen 
tft.begin(); /* TFT init */ 
// Initialize the UI 
lvgl_ui_init(); 
// Turn on the backlight
delay(50); 
lv_timer_handler(); 
digitalWrite(DISPLAY_BACKLIGHT_PIN, HIGH);
```
Hope to provide some ideas for others who encounter this issue.

**Comment by garym999** [comment]
I'm suffering a variation on the theme. I have 3 of these displays side by side used as gauges in a car. When they boot one of them gets a flash of the previous image. The others did too but playing around with timings and the order of things around tft.init sorted them out. 

I have also tried editing the INIT as above but I still get the flash 

Has there been any movement on this? I've tried reading through the display datasheet trying to see if there is a way to erase the screen buffer but I'm at the edge of my knowledge

https://www.waveshare.com/w/upload/5/5e/GC9A01A.pdf

**Comment by Basitadam** [comment]
How about displaying the first screen and then turning the backlight on?

**Reply by HerrSchwarz617** [reply]
The problem with my GC9A01 was that there was no way to control the backlight. Just look at the picture - there is no pin for it.

**Comment by garym999** [comment]
I'm actauly using a Waveshare ESP32S3 module with an integrated GC9A01 display. On the module they do expose the backlight but as part of the initialisation process it gets turned on even though I make sure its turned off before I setup the screen.

**Reply by HerrSchwarz617** [reply]
For me that's more a ESP32S3 problem as a problem with this library.
But you can ask yourself: how do you ensure that the screen is keeping off during the booting?
- Maybe it is necessary to add a resistor between the backlight pin and ground?
- Perhaps the ESP sets the backlight pin to HIGH by default when booting
- How do you control the backlight in your software? Do you have your own logic?
- Is the backlight even turning on without any software logic?

To fix the flashing disturbance or flashing the previous frame buffer flashing on boot up, you need to patch the TFT_eSPI library as I said in my initial post. Only a black screen will then be "flashed".

**Comment by garym999** [comment]
Regardless of external control, ie: turning off the backlight before TFT.INIT the backlight is turned back on by the call to TFT.INIT. Once initilaised I can adjust the brightness adjusted using my own code. 

Writing a black screen within TFT.INIT seems to have worked.
